### PR TITLE
[4.0] add fields items to contact menu

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for replace of #23366 .

### Summary of Changes

Add menu items for fields and field groups below the "Contact" menu item in the admin menu.

On update, use script.php to add these new menu items so that lft and rgt are set correctly in the (nested) menu table. 

### Testing Instructions

If you have both MySQL and PostgreSQL, please test on both if possible.

**Test 1: New installation**

1. Apply the patch on a clean 4.0-dev branch using git or merging manually, or apply it on an installation of current 4.0-dev using patchtester.
2. If used patchtester on an existing installation in step 1, delete file `configuration.php` and delete all Joomla database tables in PhpMyAdmin or PhpPgAdmin (depending on your database type).
3. Do a new installation, login to backend and check the available items below the "Contact" item in the admin menu.

Result: See section "Expected result" below.

**Test 2: Update**

1. Update a 3.9.12 or staging to current 4.0-dev plus the patch of this PR applied by using the "Upload & Update" tab of the Joomla Update Component and the update zip container previously downloaded from following URL: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-19_pr-26540.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-19_pr-26540.zip "https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-19_pr-26540.zip"). ATTENTION: When working on a git repository you should know how to clean up your branch afte this test.
2. Login to backend and check the available items below the "Contact" item in the admin menu.

Result: See section "Expected result" below.

### Expected result

you have:

- Fields
- Fieldsgroup 

below Contact menu item
Screenshot from 2019-10-10 19-39-04

### Actual result

You don't have these 2 menu items below Contact component

### Additional comments

We need to review how we deal with changes of content in nested tables when updating because issues with that are is quite hard to fix with update sql scripts. (personal consideration)

### CODE REVIEW

**Comment by richard67** [comment]
~~I think it will need some update sql script for that, too. If there is some 4.0-*.sql already existing which deals with these menu items, this one can be modified before we are in beta, otherwise or if you are not sure, make a new one. I can help with that on weekend.~~ Edit: Just after I typed that I did read statement in the description about update script not needed.

**Comment by alikon** [comment]
we need to deeply review how we provide  updates ---- if we are talking about sql
maybe i'm a dumb 
...but it has  been quite hard 4 me to to deliver this PR .... 
update path without a stored procedure  could be full of issues.... when you are talking about nested set tables, then cannot be a matter for easy/silly update  .... but... maybe i'm too silly :smile:

**Comment by brianteeman** [comment]
That's why I asked someone else to do it :)

**Comment by richard67** [comment]
Hmm, thinking more about it: If it is really ok in 3.9 and not ok in 4.0, maybe there is already some 4.0.0-*.sql update script which removes those menu items? If this is the case, then this update script has to be changed. I can check that on weekend, but tonight and tomorrow night I won't have time.

**Comment by richard67** [comment]
Well, I just checked: Seems there is no 4.0 update sql deleting those menu items, so everything should be ok, no update sql needed.

**Comment by alikon** [comment]
let me clarify...
no better
let me try to explain myself as better as I can ..
when nested set tables are in play..
imho  there are no static script that can save your ass

**Comment by richard67** [comment]
Correct.

**Comment by alikon** [comment]
you cannot guess what are the different lft,rgt values for all installation--- so
the update path should be at least careafully considered those different situation
my answer ----maybe we should consider a stored procedure......or 
i'm open to feedaback

**Comment by richard67** [comment]
but shouldn't the "rebuild" button do that after an update?

**Comment by alikon** [comment]
are you talking about menus table only ?
i'm talking about the whole update process----  i.e assets table .... etc
better ... i'm talking about an update path  where nested set table are deeply involved

**Comment by richard67** [comment]
Well I did not wanna speak against it, I only asked for my better understanding.

**Comment by alikon** [comment]
i hope i've at least instilled some grams of  curiosity :smiley:

**Comment by brianteeman** [comment]
its also why I would not have had any 4.0sql update files at all and just done them all at once when we get to beta. just creates a lot of useless files that will be deleted

**Comment by Quy** [comment]
Add a separator between `Categories` and `Fields`.

**Comment by alikon** [comment]
after a more proof review of 3.x we still need an update sql .... i was wrong

File: administrator/components/com_admin/script.php [hunk 1]
```diff
@@ -6076,4 +6076,175 @@ private function cleanJoomlaCache()
...
+				'parent_id' => $menuItem['parent_id'],
+				'client_id' => $menuItem['client_id'],
+			];
+
+			if ($tableItem->load($keys))
+			{
+				continue;
+			}
+			$newTableItem   = new \Joomla\Component\Menus\Administrator\Table\MenuTable($db);
+			// Bind the data.
```
Comment (richard67) [inline administrator/components/com_admin/script.php:6223]:
```suggestion

			// Bind the data.
```

**Review by richard67 — CHANGES_REQUESTED** [review]

**Comment by alikon** [comment]
the needed update is made now with PHP :heart_eyes:

File: administrator/components/com_admin/script.php [hunk 2]
```diff
@@ -6076,4 +6076,176 @@ private function cleanJoomlaCache()
...
+				'title'     => $menuItem['title'],
+				'parent_id' => $menuItem['parent_id'],
+				'client_id' => $menuItem['client_id'],
+			];
+
+			if ($tableItem->load($keys))
+			{
+				continue;
+			}
+			$newTableItem   = new \Joomla\Component\Menus\Administrator\Table\MenuTable($db);
```
Comment (richard67) [inline administrator/components/com_admin/script.php:6221]:
```suggestion

			$newTableItem   = new \Joomla\Component\Menus\Administrator\Table\MenuTable($db);
```

**Review by richard67 — CHANGES_REQUESTED** [review]
One last PHPCS error.

**Comment by SniperSister** [comment]
Postgres System Test is failing:
https://ci.joomla.org:444/artifacts/joomla/joomla-cms/4.0-dev/26540/system-tests/24688/InstallCest.installJoomla.postgres.fail.png

**Comment by richard67** [comment]
@SniperSister @alikon and me know about this. I've made a PR to his branch to fix it, he will merge as soon as lunch finished, I think.

**Comment by richard67** [comment]
@SniperSister Tests are passing here now.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 661f93b6993241b3ba0f27d4b3b3131f36bcf147<br /><br />Test 1 new installation - MySQL 5.7: OK.
Test 1 new installation - PostgreSQL 10.10: OK.
Test 2 update - MySQL 5.7: OK.
Test 2 update  - PostgreSQL 10.10: OK.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by richard67** [comment]
@brianteeman Could you test this, too? I've helped with updating testing instructions and making a zip package for an update test because Nicola is a bit busy with preparing for JoomlaDay Italy, so see the updated instructions for details.

File: administrator/components/com_admin/script.php [hunk 3]
```diff
@@ -6076,4 +6076,177 @@ private function cleanJoomlaCache()
...
+
+	/**
+	 * Called after any type of action
+	 *
+	 * @param   string     $action     Which action is happening (install|uninstall|discover_install|update)
+	 * @param   Installer  $installer  The class calling this method
+	 *
+	 * @return  boolean  True on success
+	 *
+	 * @since   3.7.0
```
Comment (Quy) [inline administrator/components/com_admin/script.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Comment by infograf768** [comment]
Installed on a 3.9.12

I get
Screen Shot 2019-10-14 at 12 18 20

and 
` Notice: Trying to get property 'version' of non-object in /Applications/MAMP/htdocs/trunkgitnew/administrator/components/com_joomlaupdate/Model/UpdateModel.php on line 1357`

Contacts have no fields
Screen Shot 2019-10-14 at 12 21 32
All dashboards are empty. No way to access to database page.

No way to save the administrator menu to use `Alternate preset` in order to access database page
`Save failed with the following error: Column 'publish_down' cannot be null`

used directly the link
`administrator/index.php?option=com_installer&view=database`

2 problems for database
Used `Update structure` button. 
1 problem remains and the button is no more displayed. Menus disappear.

**Comment by richard67** [comment]
@infograf768 The first message about template not available is normal after the update but should disappear and not come again after having liogged in to backend for the first time and then having navigated 1 time withint the backend.

The 2nd message should be fixed when my database null date works will be finished. It normally does not appear on a pre 8.0 MySQL with normal configuration regarding session parameters and server variables, e.g. for strict mode. Did you test this on MySQL 8?

The SQL error and empty dashboards ... maybe I've made a mistake when packing the update zip package which is linked in the testing instructions for test 2? I have to check that tonight after work (German time).

**Comment by infograf768** [comment]
> Did you test this on MySQL 8?

Nope.

**Comment by nadjak77** [comment]
I have tested this item :red_circle: unsuccessfully on 4f755bc69dc09cbec6576669b4f561cff7b74261<br /><br />Deleted configuration.php and databasetables
new installation 
applied patch
deleted browsercache and reload site
the two menue items for "fields" and "fields groups" do not appear<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by richard67** [comment]
@nadjak77 Wrong order of processing. It has to be
1. Apply patch on a 4.0-dev or alpha-12. This applies the changes to joomla.sql.
2. Remove configuration.php and clear database.
3. Run new installation. This will use the changed joomla.sql.
4. Check backend menu.

You have done the new installation and then applied the patch but then have not done new installation again, so your database did not include the changes here.

**Comment by richard67** [comment]
I've just updated the update package for Test 2 to latest nightly build of today.

Testers please follow exactly the testing instructions given in the description of this PR.

**Comment by richard67** [comment]
@nadjak77 Please change back your testing result to "not tested" in the issue tracker. The test failed because you did not follow exactly the testing instructions. And if you have enough time, please test again in the right way, that would be very appreciated. Anyway thanks for testing, even if wrong now, because we need volunteers who help with testing.

**Comment by nadjak77** [comment]
I have tested this item :white_check_mark: successfully on 4f755bc69dc09cbec6576669b4f561cff7b74261<br /><br />thank you @richard67 
I tried it again:
apply patch
delete configuration.php and database
reinstall joomla
the new items appear under contact<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by tecpromotion** [comment]
I have tested this item :white_check_mark: successfully on 4f755bc69dc09cbec6576669b4f561cff7b74261<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by uthorat** [comment]
I have tested this item :white_check_mark: successfully on 4f755bc69dc09cbec6576669b4f561cff7b74261<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26540">issues.joomla.org/tracker/joomla-cms/26540</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by astridx** [comment]
Thank you for making the custom fields usable again. 
Now I'm wondering: how should third party developers integrate custom fields in Joomla 4? Is it recommended that you also change the table `#__menu`, or should you continue to use a sub menu?

**Comment by alikon** [comment]
> Now I'm wondering: how should third party developers integrate custom fields in Joomla 4? Is it recommended that you also change the table #__menu, or should you continue to use a sub menu?

imo is matter of personal taste / consistency

**Pull request merged**